### ISSUE TITLE
HD+ channels crash the VDR

### ISSUE THREAD

**Issue by haarm** [reufer/rpihddevice#9]
I use the VDR 2.6.3 when I go to a HD+ 
channel.the VDR supports with the message.

assertion failure:ilclient.c:524:ilclient_teardown_tunnels():error == OMX_ErrorNone
This only occurs with HD+ channels.

**Comment by haarm** [comment]
With the version from here :
https://github.com/vdr-projects/vdr-plugin-rpihddevice
everything runs perfectly

**Comment by madmartin** [comment]
@haarm can you please double check that you test with the most current download from https://github.com/vdr-projects/vdr-plugin-rpihddevice ? 
I am asking because this repo has changed a few days ago to be in equal state compared to https://github.com/reufer/rpihddevice (now only a few commits behind).
Please also take a look at other branches - maybe you tested the content of the (shortly created) branch https://github.com/vdr-projects/vdr-plugin-rpihddevice/tree/MarkusEh ?!?!?

**Comment by haarm** [comment]
Will I test and report

**Comment by madmartin** [comment]
@MarkusEh @M-Reimer please take a look on this discussion please....

**Comment by haarm** [comment]
Ok, test was done quickly.
https://github.com/vdr-projects/vdr-plugin-rpihddevice/tree/MarkusEh  
does not work 

https://github.com/vdr-projects/vdr-plugin-rpihddevice 
does not work
I have probably an older version here from October .
Have also asked in the Vdr portal times, but no feedback yet

**Comment by haarm** [comment]
By the way, I have four Raspiberry ,on all of them the same

**Comment by M-Reimer** [comment]
Thanks for pinging, but I won't be able to help in any way. I don't have a Raspberry Pi set up with VDR and I don't plan to do so. Will follow the issue anyway just out of curiosity. I think you should have pinged @dr-m instead as he did the fixes which now cause problems here.

**Comment by dr-m** [comment]
I think that it would help to produce a short recording that reproduces the problem. That recording could then be used for identifying which particular commit might have caused the regression. Without such a recording, I can’t help much.

To avoid a crash while creating the recording, you can play back (or pause) a non-HD recording while creating the recording via the EPG. Once the recording is done, start to play it back and ensure that the system will crash.

All Finnish and some Estonian free-to-air DVB-T (MPEG-2) and DVB-T2 (H.264) programs that I have tried work for me, even with multilingual subtitles and multilingual audio. The only issue that I can observe is that occasionally, the OSD layer will not appear at all, that is, all menus, subtitles and so on are invisible, but VDR reacts to button presses on the remote control. Restarting VDR fixes that. It occurs rarely, maybe once in 10 to 30 startups. I have not yet debugged that.

Based on raspberrypi/firmware#1762 I would also suggest to check the firmware version. For me, that bug affects HD playback using the the proprietary MMAL interface. When I was testing different firmware versions, I did not observe broken output with the OMX interface that rpihddevice is using. But then again, I was mainly diagnosing zillevdr/vdr-plugin-softhddevice-drm#11 and can’t be sure that I tested rpihddevice HD output with each and every firmware version.

**Comment by haarm** [comment]
Sorry that my answer took a while.
I have done some tests here, I have a recording here where the problem occurs. Not always, but every other time for sure.

[00001.zip](https://github.com/reufer/rpihddevice/files/10319094/00001.zip)

**Comment by dr-m** [comment]
I tried replaying the recording several times, using cd95249f7f4f8760e7e0cd53edbceca9870d4fe6 (tagged 1.0.6) and raspberrypi/rpi-firmware@bcd35b7abb058b210b33922ba60efc1cc93f533d (Linux version 5.15.79-v7+).

I did not observe any crash here. I had GDB attached to the VDR process while attempting to replay it, at least 5 times from the beginning, and another 5 to 10 times for the last few seconds.

After the words "völlig unnötige Verletzung" were spoken, there was a scene change for a few frames, and the recording ended abruptly while most of the video frame was green.

At which timestamp would it crash for you?

One more thing that I could try is to compile VDR with `-fsanitize=address -fsanitize=undefined`. It did catch some bugs in the past.

**Comment by haarm** [comment]
The crash occurs directly when I start recording, or when I switch to live TV.

**Comment by dr-m** [comment]
Are all your Raspberry Pi the same model and running the same firmware? Did you test them all with the same SD card? Some diversity might be helpful.

Mine is a Raspberry Pi 2 model B if I remember correctly.

I started recompiling VDR 2.6.3 and cd95249f7f4f8760e7e0cd53edbceca9870d4fe6 with the [sanitizers](https://github.com/google/sanitizers/wiki/) of GCC 8.3.0. I will report back in some 10 minutes.

**Comment by dr-m** [comment]
Sorry, I overestimated the speed of my system. I had to use `make -j3` to avoid excessive swapping while compiling with both `-fsanitize=address` and `-fsanitize=undefined`. There is still no crash and no error message, nothing.

@haarm Did you reproduce a crash while playing back that recording? If so, at which timestamp?

Can you post stack traces from at least one crash?
```sh
sudo gdb -p $(pgrep vdr)
```
```gdb
handle SIGUSR1 nostop noprint
handle SIGUSR2 nostop noprint
handle SIGWAITING nostop noprint
handle SIGLWP nostop noprint
handle SIGPIPE nostop
handle SIGALRM nostop
handle SIGHUP nostop
handle SIGTERM nostop noprint
handle SIG32 nostop noprint
handle SIG33 nostop noprint
continue
```
After executing those commands, start the HD playback to provoke the crash. Once the signal is reported in GDB, type the following commands:
```gdb
set log file /tmp/log.txt
set height 0
set log on
thread apply all backtrace
quit
```
Then, attach the file `/tmp/log.txt`.

**Comment by dr-m** [comment]
I realized that the source files in `ilclient` are C, not C++. I never touched those files. They seem to be something from Broadcom.

I had only augmented `CXXFLAGS`, so those files were not covered by the sanitizers. Trying again with `CFLAGS`. On the first playback, after re-downloading the video, it almost came to a halt (lots of skipped frames, or the frame rate dropped to 1 frame per several seconds) after the words "das bedarf maximale Ressourcen" were said. But VDR did not crash. On the second, third, and fourth attempt (when the index file had already been created), the entire video played back perfectly, up to the last mostly green frame after the scene change. I have experienced this stuttering at the end of HD recordings.

@haarm How much memory are you allocating to the VideoCore GPU? In my `/boot/config.txt`, I have
```
gpu_mem=256
```
A smaller allocation will cause trouble with HD recordings. The error handling in the rpihddevice plugin could be better. I did not touch that (yet).

**Comment by dr-m** [comment]
I reviewed the changes that were merged in #6 (0c2c6aff9a1a0c6f78c87a99648833e83b5d9b64). I created a possible fix in dr-m/rpihddevice@f2f3b107837ff0e7686b0ad809fa0d9a54edb159:
```diff
diff --git a/omx.c b/omx.c
index fcb37cb..b5da3d3 100644
--- a/omx.c
+++ b/omx.c
@@ -572,6 +572,9 @@ int cOmx::DeInit(void)
 	while (Active())
 		cCondWait::SleepMs(5);
 
+	for (int i = 0; i < eNumTunnels; i++)
+		ilclient_disable_tunnel(&m_tun[i]);
+
 	ilclient_teardown_tunnels(m_tun);
 	ilclient_state_transition(m_comp, OMX_StateIdle);
 	ilclient_state_transition(m_comp, OMX_StateLoaded);
```
Shutdown does not hang or crash for me even if this patch is in, so I think that it would be good to include in any case.

This code will be invoked also during normal operation by calls to `cOmxDevice::SetPlayMode()` with `pmNone`, which I guess could be the case whenever you switch channels or start or stop playing a recording.

It looks like the `while` loop that I had added before it in 05bf186eb1606c741068737c7b3f49e8032d21b4 will prevent the crashes and hangs that I previously experienced during shutdown. Before I realized what kind of race conditions the code was suffering from, I had attempted to minimize the `ilclient` calls, but it looks like I did not put back this one because I may have believed that it is not needed.

Side note: Compiling with the default flags (no `-fsanitize`) finished in a couple of minutes with `make -j4 `.

@haarm Please check if this fixes the crashes for you.

**Comment by dr-m** [comment]
If the patch fixes the problem, then I can open a pull request, or @reufer can apply it directly with commands like this:
```sh
git pull https://github.com/dr-m/rpihddevice
git push origin
```

**Comment by haarm** [comment]
This looks very very good so far, but will test this all evening.
Everything else tomorrow,but what is still that he sometimes does not always hide the OSD.
Thanks for your effort
Carlo

**Comment by haarm** [comment]
If it does not hide the OSD only a restart helps

**Comment by haarm** [comment]
That it does not hide the OSD only occurs with HD+ channels.
But no more crash

**Comment by dr-m** [comment]
@haarm Thank you. In my opinion, this ticket should be closed by @reufer pushing my fix.

For your OSD issue, can you please file a separate ticket, perhaps with some screenshots (or photos of the screen) to illustrate how it looks like, step by step? Stack traces of the running process (attach GDB, then execute `thread apply all backtrace` and `detach`) might be helpful.

Today, I restarted VDR several dozen times, via my https://www.linuxtv.org/vdrwiki/index.php/Systemd configuration. I hoped to reproduce the "OSD does not appear" bug that I mentioned, but it never occurred. One occurrence that I remember from a few days ago started after I replayed one particular recording, but replaying that recording today did not help. I vaguely remember that at least one past occurrence was so that the OSD failed to display straight from the beginning.

**Comment by haarm** [comment]
I watched live tv all evening yesterday, and two recordings.all without a crash.
For the OSD problem, I'll open a new ticket.

**Comment by reufer** [comment]
> If the patch fixes the problem, then I can open a pull request, or @reufer can apply it directly with commands like this:
> 
> ```shell
> git pull https://github.com/dr-m/rpihddevice
> git push origin
> ```

Thanks a lot @dr-m for investigating and providing a fix! I'm happy to merge if you could do a PR. I'm not having my development infrastructure at hand currently, so I'm sorry for my limited support right now...

**Comment by dr-m** [comment]
> > If the patch fixes the problem, then I can open a pull request, or @reufer can apply it directly with commands like this:
> > ```shell
> > git pull https://github.com/dr-m/rpihddevice
> > git push origin
> > ```
> 
> Thanks a lot @dr-m for investigating and providing a fix! I'm happy to merge if you could do a PR. I'm not having my development infrastructure at hand currently, so I'm sorry for my limited support right now...

Very well, I just filed pull request #11.
